### PR TITLE
CC-1257: Add hadolint to `course-sdk lint`

### PR DESCRIPTION
**Pull request by ryan-gang**

### CODE REVIEW

File: commands/lint.ts [hunk 1]
```diff
@@ -73,7 +82,13 @@ export default class LintCommand extends BaseCommand {
 
   async lintRustFiles() {
     const dockerShellCommandExecutor = await this.dockerShellCommandExecutor("rust-tools");
-    await dockerShellCommandExecutor.exec(`find . -name '*.rs' | xargs rustfmt --edition "2021" --check`);
+    await dockerShellCommandExecutor.exec(`find . -name '*.rs' -exec rustfmt --edition "2021" --check -- {} +`);
+  }
+
+  async lintDockerFiles() {
+    const dockerShellCommandExecutor = await this.dockerShellCommandExecutor("docker-tools");
+    await dockerShellCommandExecutor.exec(`sed -i "/^COPY /s/--exclude=[^ ]*//g" dockerfiles/*.Dockerfile`);
```
Comment (rohitpaulk) [inline commands/lint.ts]: @ryan-gang will this mutate files in the directory? If so we should probably copy it elsewhere?

File: lib/docker-shell-command-executor.ts [hunk 2]
```diff
@@ -32,7 +33,10 @@ export default class DockerShellCommandExecutor {
   }
 
   async exec(command: string) {
-    await exec(`docker run --rm -v ${this.workingDirectory}:/workdir -w /workdir course-sdk-${this.dockerfileType} ${command}`);
+    let quotedCommand = command.replace(/\\/g, '\\\\')
+            .replace(/"/g, '\\"')
+            .replace(/'/g, "\\'");
```
Comment (rohitpaulk) [inline lib/docker-shell-command-executor.ts]: @ryan-gang why do single quotes need to be replaced when we're surrounding a string with double quotes?

File: lib/docker-shell-command-executor.ts [hunk 3]
```diff
@@ -32,7 +33,10 @@ export default class DockerShellCommandExecutor {
   }
 
   async exec(command: string) {
-    await exec(`docker run --rm -v ${this.workingDirectory}:/workdir -w /workdir course-sdk-${this.dockerfileType} ${command}`);
+    let quotedCommand = command.replace(/\\/g, '\\\\')
```
Comment (rohitpaulk) [inline lib/docker-shell-command-executor.ts]: What does this do?

**Review by rohitpaulk — CHANGES_REQUESTED** [review]
Added q

File: lib/docker-shell-command-executor.ts (hunk 3 above)
Comment (ryan-gang) [reply lib/docker-shell-command-executor.ts]: Globally replaces backslashes, with 2 backslashes, ie their escaped version.

File: commands/lint.ts [hunk 4]
```diff
@@ -54,6 +54,22 @@ export default class LintCommand extends BaseCommand {
...
+
+    await this.buildToolsImage("docker-tools");
+
+    console.log("");
+    console.log("Linting Docker files...")
+    console.log("");
+    const tmpDirectory = "dockerfiles-tmp/";
+    try {
+      await this.lintDockerFiles(tmpDirectory);
+      await this.cleanupAfterLintDockerFiles(tmpDirectory);
```
Comment (rohitpaulk) [inline commands/lint.ts]: @ryan-gang all of this logic (tmp, cleanup etc.) should just be inside `lintDockerfiles` - don't see why we're pulling it out here

File: commands/lint.ts [hunk 5]
```diff
@@ -73,7 +89,19 @@ export default class LintCommand extends BaseCommand {
...
+  async lintDockerFiles(tmpDirectory: string) {
+    const dockerShellCommandExecutor = await this.dockerShellCommandExecutor("docker-tools");
+    await dockerShellCommandExecutor.exec(`cp -r dockerfiles/ ${tmpDirectory}/`);
+    await dockerShellCommandExecutor.exec(`sed -i "/^COPY /s/--exclude=[^ ]*//g" ${tmpDirectory}/*.Dockerfile`);
+    await dockerShellCommandExecutor.exec(`hadolint --ignore DL3059 ${tmpDirectory}/*.Dockerfile`);
+  }
+  
+  async cleanupAfterLintDockerFiles(tmpDirectory: string) {
+    const dockerShellCommandExecutor = await this.dockerShellCommandExecutor("docker-tools");
+    await dockerShellCommandExecutor.exec(`rm -rf ${tmpDirectory}`);
```
Comment (rohitpaulk) [inline commands/lint.ts]: We don't need dockerShellCommandExecutor to remove files here btw, regular `ShellCommandExecutor` would do

**Review by rohitpaulk — APPROVED** [review]

File: lib/docker-shell-command-executor.ts (hunk 3 above)
Comment (rohitpaulk) [reply lib/docker-shell-command-executor.ts]: Ah, okay makes sense

**Pull request merged**